optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,422 @@
+    ) -> Tuple[Optional[float], Optional[float], Optional[float]]:
+
+        # We calculate low and high.
+        if isinstance(distribution, distributions.UniformDistribution):
+            low = distribution.low
+            high = distribution.high
+            q = None
+        elif isinstance(distribution, distributions.LogUniformDistribution):
+            low = np.log(distribution.low)
+            high = np.log(distribution.high)
+            q = None
+        elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+            q = distribution.q
+            low = distribution.low - 0.5 * q
+            high = distribution.high + 0.5 * q
+        elif isinstance(distribution, distributions.IntUniformDistribution):
+            q = distribution.step
+            low = distribution.low - 0.5 * q
+            high = distribution.high + 0.5 * q
+        elif isinstance(distribution, distributions.IntLogUniformDistribution):
+            low = np.log(distribution.low - 0.5)
+            high = np.log(distribution.high + 0.5)
+            q = None
+        else:
+            distribution_list = [
+                distributions.UniformDistribution.__name__,
+                distributions.LogUniformDistribution.__name__,
+                distributions.DiscreteUniformDistribution.__name__,
+                distributions.IntUniformDistribution.__name__,
+                distributions.IntLogUniformDistribution.__name__,
+                distributions.CategoricalDistribution.__name__,
+            ]
+            raise NotImplementedError(
+                "The distribution {} is not implemented. "
+                "The parameter distribution should be one of the {}".format(
+                    distribution, distribution_list
+                )
+            )
+
+        assert low < high
+
+        return low, high, q
+
+    def _transform_to_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.log(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.IntLogUniformDistribution):
+                transformed[param_name] = np.log(samples)
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                transformed[param_name] = samples
```

Comment:
Hmm... I'm not sure the current one helps the readability for developers. The newbies for this multivariate TPE don't know that the same conditional statements are used in `_transform_to_uniform`, `_transform_from_uniform` and `_calculate_parzen_bounds`. Therefore, to reduce the amount of information at first glance of newbies, I think my suggested simplification is useful. 

I would like to know @hvy's opinion. What do you think?

Reply:
Purely speaking about style and readability, I understand @HideakiImamura 's suggestion is more consistent with what we already have, which might be good in itself. From a readability point of view, @kstoneriv3 's original code contains fewer lines and less indentation, so I personally find it quite pleasing. 

Another idea is the following. I think this makes the code very clear in what it does. 

```python
def _transform_to_uniform(
    self, multivariate_samples: Dict[str, np.ndarray]
) -> Dict[str, np.ndarray]:

    transformed = {}
    for param_name, samples in multivariate_samples.items():
        distribution = self._search_space[param_name]

        assert distribution in distributions.DISTRIBUTION_CLASSES
        if isinstance(distribution, (LogUniformDistribution, IntLogUniformDistribution)):
            samples = np.log(samples)

        transformed[param_name] = samples
    return transformed
```

As for the assertion, I suggest we check all the distribution types in `__init__` and assert in following usages.

Reply:
OK, I'll adopt @hvy 's idea.
As for assertion, your statements were a bit unclear for me. Did you mean to assert all distributions in the `search_space` is in `DISTRIBUTION_CLASSES` in `__init__` ?

Reply:
I think @hvy means that we should raise a `NotImplementedError` as done in `TPESampler.sample_independent`.

Reply:
Thanks @HideakiImamura for the clarification. Yes, I meant to raise in `_MultivariateParzenEstimator.__init__` after checking. Then, we can assume in following usages within the class. But I now think that it might be better to explicitly list all distribution instead of using `DISTRIBUTION_CLASSES` as this container may change.

Reply:
@hvy @HideakiImamura  This kind of situation is happening in both sampler.py and multivariate_parzen_estimator.py. I would like to take a unified approach to them. Which of the following do you think is the best approach?
- use `DISTRIBUTION_CLASSES`
- use `_DISTRIBUTION_CLASSES` (private variable)
- use an explicit tuple or list of all possible distributions.

Reply:
Good question. If a new distribution is introduced, either by a user or by Optuna (appended to `DISTRIBUTION_CLASSES`), my understand is that it's more user friendly to fall back to random sampling with a warning rather than raising an error. In this case, the second (or third) option sounds reasonable creating a private variable and reusing it throughout, including in `infer_relative_search_space`.

Reply:
OK, I will go for the second option.